## Disciplina: Aprendizado Estatístico

### Estudo Dirigido 2

- Conforme discutido em aula, há muitas maneiras de selecionar o modelo. Veremos métodos de seleção forward e backward utilizando o AIC.
---
#### Exercício 1: Tratamento dos dados

- Vamos carregar os dados da temporada da Major League Baseball de 2011.

In [1]:
# Baixar e carregar os dados
download.file("http://www.openintro.org/stat/data/mlb11.RData", destfile = "mlb11.RData")
load("mlb11.RData")

# Visualizar os dados
str(mlb11)

'data.frame':	30 obs. of  12 variables:
 $ team        : Factor w/ 30 levels "Arizona Diamondbacks",..: 28 4 10 13 26 18 19 16 9 12 ...
 $ runs        : int  855 875 787 730 762 718 867 721 735 615 ...
 $ at_bats     : int  5659 5710 5563 5672 5532 5600 5518 5447 5544 5598 ...
 $ hits        : int  1599 1600 1540 1560 1513 1477 1452 1422 1429 1442 ...
 $ homeruns    : int  210 203 169 129 162 108 222 185 163 95 ...
 $ bat_avg     : num  0.283 0.28 0.277 0.275 0.273 0.264 0.263 0.261 0.258 0.258 ...
 $ strikeouts  : int  930 1108 1143 1006 978 1085 1138 1083 1201 1164 ...
 $ stolen_bases: int  143 102 49 153 57 130 147 94 118 118 ...
 $ wins        : int  96 90 95 71 90 77 97 96 73 56 ...
 $ new_onbase  : num  0.34 0.349 0.34 0.329 0.341 0.335 0.343 0.325 0.329 0.311 ...
 $ new_slug    : num  0.46 0.461 0.434 0.415 0.425 0.391 0.444 0.425 0.41 0.374 ...
 $ new_obs     : num  0.8 0.81 0.773 0.744 0.766 0.725 0.788 0.75 0.739 0.684 ...


- Além das runs marcados, há sete variáveis tradicionalmente usadas no conjunto de dados: at-bats, hits, home runs, batting average, strikeouts, stolen bases, e wins. Existem também três variáveis menos tradicionais: on-base percentage, slugging percentage, e on-base plus slugging. Para a primeira parte da análise consideraremos as sete variáveis tradicionais. No final deste estudo dirigido , você terá trabalhará sozinho com as variáveis mais recentes. Também gostarı́amos de modificar os dados para facilitar o trabalho durante a seleção do modelo. Removemos a variável "team"do conjunto de dados e armazenamos a versão atualizada em "mlb11_wins".

In [3]:
# Remover a coluna 'team' (1ª coluna)
mlb11_wins <- mlb11[ , -1]

- Como as vitórias não são uma estatı́stica de nı́vel de jogador ( pelo menos para não-arremessadores) não queremos usá-la ao prever runs. Portanto, criaremos outro conjunto de dados modificado para utilizar ao tentar encontrar o melhor modelo para prever o número total de (runs )corridas de uma equipe durante uma temporada.

In [4]:
# Criar conjunto para análise de "runs"
mlb11_runs <- mlb11_wins[ , -8]  # Remover a coluna 'wins'

#### Exercício 2: Previsão de execuções com seleção retroativa (Backward):

In [7]:
# Modelo completo para prever 'runs'
runs_full <- lm(runs ~ ., data = mlb11_runs)
summary(runs_full)


Call:
lm(formula = runs ~ ., data = mlb11_runs)

Residuals:
    Min      1Q  Median      3Q     Max 
-40.062 -10.675   0.089   9.404  28.293 

Coefficients:
               Estimate Std. Error t value Pr(>|t|)  
(Intercept)   2.448e+03  2.775e+03   0.882   0.3882  
at_bats      -5.565e-01  4.975e-01  -1.118   0.2766  
hits          2.578e+00  1.934e+00   1.333   0.1975  
homeruns      3.937e-01  4.537e-01   0.868   0.3957  
bat_avg      -1.506e+04  1.081e+04  -1.393   0.1790  
strikeouts   -2.835e-02  4.828e-02  -0.587   0.5637  
stolen_bases  3.209e-01  1.343e-01   2.389   0.0269 *
new_onbase   -3.551e+03  6.888e+03  -0.515   0.6119  
new_slug     -6.049e+03  6.947e+03  -0.871   0.3942  
new_obs       6.972e+03  6.850e+03   1.018   0.3209  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 19.1 on 20 degrees of freedom
Multiple R-squared:  0.963,	Adjusted R-squared:  0.9464 
F-statistic: 57.86 on 9 and 20 DF,  p-value: 2.327e-12


- Agora que temos um modelo completo definido, podemos proceder à seleção reversa do modelo. A função step() em R torna extremamente fácil o uso do AIC (Critério de Informação de Akiake) para seleção de modelo. AIC aplica uma penalidade aos modelos que usam mais variáveis preditoras. Execute o código a seguir para determinar o melhor modelo para prever as corridas de uma equipe em uma temporada usando seleção retroativa com AIC como critério (observe que um AIC mais baixo indica um modelo melhor).

In [8]:
# Seleção backward usando AIC
runs_backAIC <- step(runs_full, direction = "backward")
summary(runs_backAIC)

Start:  AIC=184.82
runs ~ at_bats + hits + homeruns + bat_avg + strikeouts + stolen_bases + 
    new_onbase + new_slug + new_obs

               Df Sum of Sq    RSS    AIC
- new_onbase    1     96.94 7394.1 183.22
- strikeouts    1    125.80 7423.0 183.33
- homeruns      1    274.82 7572.0 183.93
- new_slug      1    276.65 7573.8 183.94
- new_obs       1    377.98 7675.1 184.34
- at_bats       1    456.40 7753.6 184.64
<none>                      7297.2 184.82
- hits          1    648.26 7945.4 185.38
- bat_avg       1    707.75 8004.9 185.60
- stolen_bases  1   2081.88 9379.0 190.35

Step:  AIC=183.22
runs ~ at_bats + hits + homeruns + bat_avg + strikeouts + stolen_bases + 
    new_slug + new_obs

               Df Sum of Sq     RSS    AIC
- strikeouts    1     140.1  7534.2 181.78
- homeruns      1     311.1  7705.2 182.45
<none>                       7394.1 183.22
- at_bats       1     593.0  7987.1 183.53
- hits          1     803.8  8197.9 184.31
- bat_avg       1     882.2  8276


Call:
lm(formula = runs ~ hits + bat_avg + stolen_bases + new_slug + 
    new_obs, data = mlb11_runs)

Residuals:
    Min      1Q  Median      3Q     Max 
-47.974 -10.020  -2.483  14.455  35.314 

Coefficients:
               Estimate Std. Error t value Pr(>|t|)    
(Intercept)   -685.4474   116.3816  -5.890 4.47e-06 ***
hits             0.4367     0.2659   1.642   0.1136    
bat_avg      -3986.8871  1982.7086  -2.011   0.0557 .  
stolen_bases     0.2704     0.1178   2.295   0.0308 *  
new_slug     -1606.1502   851.8902  -1.885   0.0715 .  
new_obs       3325.3137   675.1944   4.925 5.03e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 18.53 on 24 degrees of freedom
Multiple R-squared:  0.9582,	Adjusted R-squared:  0.9495 
F-statistic: 110.2 on 5 and 24 DF,  p-value: 9.595e-16


O modelo final inclui as variáveis hits, bat_avg, stolen_bases, new_slug, e new_obs, e não houve mais
remoções. O AIC final é igual a 180.46. Com isso, parte-se para as respostas de cada uma das seis perguntas:

1) Quantas etapas a seleção retroativa usando AIC conduziu antes de selecionar um modelo?**

Foram realizadas quatro etapas antes de chegar ao modelo final.

2) Qual variável foi a primeira a ser removida?

A variável new_onbase foi a primeira a ser removida.

3) Quais variáveis acabaram no modelo final?

As variáveis no modelo final são hits, bat_avg, stolen_bases, new_slug, e new_obs.

4) Quantos parâmetros estão sendo estimados neste modelo final?

Há cinco parâmetros (um para cada variável preditora), além da interceptação (ou intercepto). Portanto, seis parâmetros no total.

5) Quantos dos parâmetros neste modelo final são significativamente diferentes de 0 no nı́vel 0.05?

Os p-valores indicam a probabilidade de observar um valor do coeficiente tão extremo quanto o observado, assumindo que a hipótese nula de que o coeficiente é igual a zero seja verdadeira. Portanto, p-valores menores que o nível de significância escolhido (neste caso, 0.05) são considerados estatisticamente significativos. Os coeficientes para interceptação, stolen_bases, new_obs têm p-valor abaixo de 0.05, indicando significância estatística. Os outros coeficientes não são estatisticamente significativos a um nível de 0.05. Todos os mencionados neste parágrafo são diferentes de 0.

6) Este modelo final tem melhor AIC do que o modelo completo para execuções?

Sim, o AIC final de 180.46 é menor do que o AIC do modelo completo igual a 184.82, indicando que o modelo final tem um ajuste melhor de acordo com o critério AIC.

#### Exercício 3: Previsão de execuções com seleção direta (Forward)

- O primeiro passo forward é configurar um modelo nulo/base a partir do qual construir. Este modelo poderia incluir variáveis que os pesquisadores estipulam que um modelo deve ter por razões teóricas. Essas variáveis não existem em nosso caso, o que significa que nosso modelo nulo terá apenas a interceptação. Também devemos especificar o modelo completo para que o procedimento saiba quais modelos tentar. Observe que o modelo completo será igual ao da seleção retroativa.

In [11]:
# Modelo nulo
runs_null <- lm(runs ~ 1, data = mlb11_runs)
summary(runs_null)


Call:
lm(formula = runs ~ 1, data = mlb11_runs)

Residuals:
   Min     1Q Median     3Q    Max 
-137.6  -64.6   11.9   40.4  181.4 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   693.60      15.06   46.06   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 82.48 on 29 degrees of freedom


- Agora que temos um modelo nulo definido, podemos proceder à seleção direta do modelo. Mais uma vez usaremos a função step() em R para usar AIC (Critério de Informação de Akiake) para seleção de modelo - lembre-se de que AIC mais baixo indica um modelo melhor.

In [12]:
# Seleção forward
runs_forwardAIC <- step(runs_null, direction = "forward", scope = formula(runs_full))
summary(runs_forwardAIC)

Start:  AIC=265.74
runs ~ 1

               Df Sum of Sq    RSS    AIC
+ new_obs       1    184444  12838 185.77
+ new_slug      1    176936  20346 199.58
+ new_onbase    1    167513  29769 211.00
+ bat_avg       1    129432  67850 235.72
+ hits          1    126642  70639 236.92
+ homeruns      1    123609  73672 238.19
+ at_bats       1     73559 123722 253.74
+ strikeouts    1     33411 163870 262.17
<none>                      197281 265.74
+ stolen_bases  1       575 196706 267.65

Step:  AIC=185.77
runs ~ new_obs

               Df Sum of Sq   RSS    AIC
+ stolen_bases  1   2563.86 10274 181.09
<none>                      12838 185.77
+ bat_avg       1    642.46 12195 186.23
+ new_slug      1    526.05 12312 186.51
+ new_onbase    1    374.87 12463 186.88
+ hits          1    329.49 12508 186.99
+ strikeouts    1     77.46 12760 187.59
+ homeruns      1     34.21 12803 187.69
+ at_bats       1      3.97 12834 187.76

Step:  AIC=181.08
runs ~ new_obs + stolen_bases

             D


Call:
lm(formula = runs ~ new_obs + stolen_bases, data = mlb11_runs)

Residuals:
    Min      1Q  Median      3Q     Max 
-38.762 -13.146   0.160   9.837  33.519 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -731.1710    65.0951 -11.232 1.11e-11 ***
new_obs      1933.3858    87.3456  22.135  < 2e-16 ***
stolen_bases    0.3154     0.1215   2.596   0.0151 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 19.51 on 27 degrees of freedom
Multiple R-squared:  0.9479,	Adjusted R-squared:  0.9441 
F-statistic: 245.7 on 2 and 27 DF,  p-value: < 2.2e-16


- Usando todas as variáveis disponı́veis em nosso conjunto de dados, conduza a seleção retroativa usando AIC para selecionar o melhor modelo para prever vitórias (wins) de uma equipe em uma única temporada. Dicas: certifique-se de usar o conjunto de dados "mlb11_wins".

In [13]:
wins_null <- lm(wins ~ 1, data = mlb11_wins)
summary(wins_null)
wins_full <- lm(wins ~ ., data = mlb11_wins)
wins_forwardAIC <- step(wins_null, direction = "forward", scope = formula(wins_full))
summary(wins_forwardAIC)


Call:
lm(formula = wins ~ 1, data = mlb11_wins)

Residuals:
     Min       1Q   Median       3Q      Max 
-24.9667  -8.9667  -0.9667   9.0333  21.0333 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   80.967      2.084   38.85   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 11.42 on 29 degrees of freedom


Start:  AIC=147.08
wins ~ 1

               Df Sum of Sq    RSS    AIC
+ homeruns      1   1649.18 2129.8 131.88
+ new_slug      1   1425.43 2353.5 134.87
+ new_obs       1   1415.74 2363.2 135.00
+ runs          1   1364.10 2414.9 135.65
+ new_onbase    1   1152.63 2626.3 138.16
+ bat_avg       1    465.02 3313.9 145.14
+ hits          1    334.82 3444.1 146.30
+ strikeouts    1    297.81 3481.2 146.62
<none>                      3779.0 147.08
+ stolen_bases  1     15.77 3763.2 148.96
+ at_bats       1     14.60 3764.4 148.96

Step:  AIC=131.88
wins ~ homeruns

               Df Sum of Sq    RSS    AIC
+ at_bats       1   153.276 1976.5 131.64
<none>                      2129.8 131.88
+ new_onbase    1   128.339 2001.5 132.01
+ strikeouts    1   109.761 2020.0 132.29
+ new_obs       1    64.573 2065.2 132.95
+ runs          1    61.401 2068.4 133.00
+ new_slug      1    28.850 2100.9 133.47
+ bat_avg       1     7.508 2122.3 133.77
+ hits          1     0.870 2128.9 133.87
+ stolen_ba


Call:
lm(formula = wins ~ homeruns + at_bats + new_onbase + strikeouts, 
    data = mlb11_wins)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.1124  -3.9547  -0.5518   2.6519  20.8810 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)   
(Intercept) 357.69334  123.61509   2.894  0.00779 **
homeruns      0.18240    0.05126   3.558  0.00153 **
at_bats      -0.06915    0.02274  -3.040  0.00548 **
new_onbase  325.75679  164.22009   1.984  0.05838 . 
strikeouts   -0.02336    0.01570  -1.487  0.14940   
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 7.584 on 25 degrees of freedom
Multiple R-squared:  0.6195,	Adjusted R-squared:  0.5586 
F-statistic: 10.17 on 4 and 25 DF,  p-value: 4.971e-05


- Quantas variáveis estão sendo usadas para prever vitórias no modelo completo?

Há quatro variáveis sendo usadas para prever as vitórias no modelo final. Além disso, há quatro parâmetros (um para cada variável preditora), além da interceptação (ou intercepto). Portanto, cinco parâmetros no total.

- Quais variáveis estão incluı́das neste modelo final?

As variáveis no modelo final são homeruns, at_bats, new_onbase, e strikeouts. Os coeficientes para interceptação, homeruns e at_bats têm p-valor abaixo de 0.05, indicando significância estatística. Os outros coeficientes não são estatisticamente significativos a um nível de 0.05. Todos os mencionados neste parágrafo são diferentes de 0.